### PR TITLE
Define getTempRet0/setTempRet0 in native code

### PR DESCRIPTION
**Pull request by sbc100**
See #7273

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -562,6 +562,10 @@ def get_binaryen_passes():
     if settings.LEGALIZE_JS_FFI:
       # legalize it again now, as the instrumentation may need it
       passes += ['--legalize-js-interface']
+      if settings.RELOCATABLE:
+        passes += ['--pass-arg=legalize-js-interface-export-originals']
+      if not settings.SIDE_MODULE:
+        passes += ['--pass-arg=legalize-js-interface-exported-helpers']
   if settings.EMULATE_FUNCTION_POINTER_CASTS:
```
Comment (kripken) [inline emcc.py:569]: How does this work for a side module - is it exporting the helpers from the side module to the outside? I'd expect the reverse so I'm not sure I follow.

File: emscripten.py [hunk 2]
```diff
@@ -497,15 +496,22 @@ def finalize_wasm(infile, outfile, memfile):
...
   else:
-    args.append('--no-legalize-javascript-ffi')
+    if settings.LEGALIZE_JS_FFI:
+      # When we dynamically link our JS loader adds functions from wasm modules to
+      # the table. It must add the original versions of them, not legalized ones,
+      # so that indirect calls have the right type, so export those.
+      if settings.RELOCATABLE:
+        args.append('--pass-arg=legalize-js-interface-export-originals')
+      if not settings.SIDE_MODULE:
+        args.append('--pass-arg=legalize-js-interface-exported-helpers')
```
Comment (kripken) [inline emscripten.py:511]: These 4 lines are duplicated in `emcc.py` - maybe a helper?

File: tools/system_libs.py [hunk 3]
```diff
@@ -311,7 +311,8 @@ def build_objects(self, build_dir):
       else:
         cmd = [shared.EMXX]
 
-      cmd += cflags
+      if ext != '.s':
+        cmd += cflags
```
Comment (kripken) [inline tools/system_libs.py:315]: What is the difference between `.S` and `.s` files? (I guess that difference explains the cflags handling difference?)

File: emcc.py (hunk 1 above)
Comment (sbc100) [reply emcc.py:569]: Side modules remain unchanged in that they continue to import these helpers.  The helpers are defined in compiler-rt which is always/only linked into the main module.  The main module exports them and side modules import them.

File: tools/system_libs.py (hunk 3 above)
Comment (sbc100) [reply tools/system_libs.py:315]: `.S` files go through the pre-processor and so flags like `-D` and `-I` are valid. `-s` files are fed directly the assembler.

Comment (sbc100) [reply tools/system_libs.py:320]: Added a comment

File: emcc.py [hunk 4]
```diff
@@ -562,6 +562,10 @@ def get_binaryen_passes():
     if settings.LEGALIZE_JS_FFI:
       # legalize it again now, as the instrumentation may need it
       passes += ['--legalize-js-interface']
+      if settings.RELOCATABLE:
+        passes += ['--pass-arg=legalize-js-interface-export-originals']
+      if not settings.SIDE_MODULE:
+        passes += ['--pass-arg=legalize-js-interface-exported-helpers']
```
[the remark this answers is not in the document]
Comment (kripken) [reply emcc.py]: I see, then maybe comments here could help, saying what you just did here in the comment. That wasn't obvious to me from the names, sorry.

File: emcc.py [hunk 5]
```diff
@@ -562,6 +562,7 @@ def get_binaryen_passes():
     if settings.LEGALIZE_JS_FFI:
       # legalize it again now, as the instrumentation may need it
       passes += ['--legalize-js-interface']
+      passes += building.js_legalizeation_pass_flags()
```
Comment (kripken) [inline emcc.py:565]:
```suggestion
      passes += building.js_legalization_pass_flags()
```

**Review by kripken — APPROVED** [review]
lgtm % comments

**Pull request merged**